gadget/gadget.go

```diff
@@ -299,41 +329,99 @@ func validateVolume(name string, vol *Volume) error {
 		return fmt.Errorf("invalid schema %q", vol.Schema)
 	}
 
-	structureNames := make(map[string]bool, len(vol.Structure))
+	// named structures, for cross-referencing relative offset-write names
+	knownStructures := make(map[string]*PositionedStructure, len(vol.Structure))
+	// for validating structure overlap
+	structures := make([]PositionedStructure, len(vol.Structure))
+
+	previousEnd := Size(0)
 	for idx, s := range vol.Structure {
 		if err := validateVolumeStructure(&s, vol); err != nil {
-			return fmt.Errorf("invalid structure %s: %v", fmtIndexAndName(idx, s.Name), err)
+			return fmt.Errorf("invalid structure %v: %v", fmtIndexAndName(idx, s.Name), err)
+		}
+		start := s.Offset
+		if start == 0 {
+			start = previousEnd
+		}
```

Comment:
Since structures are a sequence I believe this is not correct.

Consider this example.
- structure A, size 10, offset 5
- structure B, size 5, offset 0

Reading this literally I would expect the structures to be back to back, with B being first, followed by A.

The code will make the structures be arranged A, B, with B starting at offset 15.

Reply:
Pushed an update.